### ISSUE TITLE
[soundfiler] request: extract loop regions from .aiff 

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#586]
.aiff files contain information on loop regions that would be useful to create players in Pd that can load this files and play them with a sustained loop. the [soundfiler] object seems like the candidate for such a task.

For reference, Max has this [info~] object that gets such loop information from .aiff files, see: [https://docs.cycling74.com/max7/maxobject/info~](https://docs.cycling74.com/max7/maxobject/info~)

I figure we could add extra floats in the list that's currently being sent to the right outlet of [soundfiler] in the case we have such loop regions marked in the file. Or maybe yet another outlet for them?

By the way, there's been a discussion in the dev list about [soundfiler] also supporting .caf files, which also encode such information. In this case, this request can also be extended to .caf files as well.

cheers

**Comment by umlaeute** [comment]
CAF files support arbitrary user data chunks.
WAV files support quite a number of non-audio chunks.

i think `[soundfiler]` should be kept simple and only support information compatible with all supported formats (NeXT `.snd`!).

For extracting non-audio data from a soundfile, we probably need another object... `[soundfile_info]` or somesuch.

**Comment by porres** [comment]
Does this mean sample rate, number of channels and the like (currently output bu [soundfiler]) are not 'non-audio data'?

**Comment by Lucarda** [comment]
Yes. 
That data is extracted from the file "header".
Here is a WAV header explained: http://soundfile.sapp.org/doc/WaveFormat/

I guess you can make an [info~] for .aiff and get the values you are asking. You can inspect https://git.iem.at/pd/iemlib/blob/master/src/soundfile_info.c
The .aiff header is similar to .wav but iirc is not fixed size.

**Comment by porres** [comment]
With the help of AI, I was able to code an object that extracts loop and 'INST' (instrument) information from analyzing '.aif' file's header in a similar fashion than [soundfiler]. I am including such an object in ELSE, but I may simply use 'libsndfile' to get this and much more information. So I did this 'the hardcore' way so far thinking about making a PR for Pd. One way or another, I think I'll be adding this new external at one place or another so people can test and check the code.

In the case of adding this to Pd, we'd just need to expand d_soundfile_aiff.c, which already handles the header information.

> For extracting non-audio data from a soundfile, we probably need another object... [soundfile_info] or somesuch.

I was thinking about this @umlaeute - adding all sorts of metadata output could make [soundfiler] more complex than it already is. Nonetheless, d_soundfile_aiff.c would still be the place to add this. Not sure about the design yet. 

> Yes.
> That data is extracted from the file "header".

Sure @Lucarda 

> Does this mean sample rate, number of channels
> and the like (currently output by [soundfiler]) are not 'non-audio data'?

I was asking this but thinking that this was information and not "audio data", by audio data I was thinking about the actual audio samples. I now understand this would be "audio related information", like the sample rate and stuff, while other metadata would be information about the file, not audio.

Anyway, here's my object prototype

<img width="722" alt="Image" src="https://github.com/user-attachments/assets/00a74fc9-4726-4fa6-be31-957f970c7d18" />

For the time being, the object is in here https://github.com/porres/pd-else/blob/master/Source/Control/aiffinfo.c

------

So, I needed this to extract loop regions from aif files, so I could build samplers that respect the loop and I have a bunch of aif files. This is all I care. I am not bothering extracting other metadata info and maybe we shouldn't. I think .caf also has the same specifications and informations to extract, so next step would be to tackle that. Not sure about WAV though (I guess not).

Hence, it'd be a lot of trouble to add another outlet just for that or another object just for that... tell me if maybe I should just keep this in an external perhaps. 

What do you say? Also tagging @danomatika on this.

**Comment by porres** [comment]
> Not sure about WAV

yeah, it does have sample info as well, so .wav, .aiff and .caf do have this, while I'm betting next format does not, neither text files of course, which [soundfiler] also manages.

> i think [soundfiler] should be kept simple and only support information compatible with all supported formats (NeXT .snd!).

I'm really starting to agree with this now, so a new [soundfileinfo] object for these 3 formats is what we could have

**Comment by danomatika** [comment]
I *think* I had a branch that parsed that info when reading the header. It would output from soundfilers right outlet. If I remember correctly, I almost had this working for WAV and AIFF but something became too complicated so I put on the side in favor of getting the sound file rewrite done.enohp ym morf tnes-----------Dan Wilcoxdanomatika.comrobotcowboy.comOn Feb 1, 2025, at 7:31 PM, porres ***@***.***> wrote:﻿

Not sure about WAV

yeah, it does have sample info as well, so .wav, .aiff and .caf do have this, while I'm betting next format does not, neither text files of course, which [soundfiler] also manages.

i think [soundfiler] should be kept simple and only support information compatible with all supported formats (NeXT .snd!).

I'm really starting to agree with this now, so a new [soundfileinfo] object for these 3 formats is what we could have

**Comment by porres** [comment]
I'm deciding to just close this issue, seems like a very niche feature and now I can provide it as an external if anyone really needs it. It's not that crucial or required, not to mention all the work. In the end, if someone really really really needs to do something like this in Vanilla for a particular project, it is possible to parse the information with [file] :)
